src/xref2/tools.ml

```diff
@@ -1270,28 +1281,35 @@ and fragmap :
 
   let sub = List.fold_right sub_of_removed removed Subst.identity in
 
-  let map_items subfn items =
+  let map_items items =
     (* Invalidate resolved paths containing substituted idents - See the `With11`
        test for an example of why this is necessary *)
     let sub_of_substituted x sub =
       let x = (x :> Ident.path_module) in
-      subfn x sub
+      (if mark_substituted then Subst.add_module_substitution x sub else sub)
+      |> Subst.path_invalidate_module x
+      |> Subst.mto_invalidate_module x
     in
+
     let substituted_sub =
       List.fold_right sub_of_substituted subbed_modules Subst.identity
     in
     (* Need to call `apply_sig_map` directly as we're substituting for an item
        that's declared within the signature *)
-    let sg = Subst.apply_sig_map substituted_sub items [] in
+    let sg = Subst.apply_sig_map substituted_sub items [] false in
     (* Finished marking substituted stuff *)
     sg.items
   in
 
-  let items = map_items Subst.add_module_substitution items in
+  let items = map_items items in
 
   let res =
     Subst.signature sub
-      { Component.Signature.items; removed = removed @ sg.removed }
+      {
+        Component.Signature.items;
+        removed = removed @ sg.removed;
+        compiled = false;
```

Comment:
`sg.compiled` ?

Reply:
not without a bit more logic - there may be some circumstances under which we don't need to recompile (if there are only type equalities added?) but most we will do.